proposals/csharp-10.0/lambda-improvements.md

```diff
@@ -0,0 +1,200 @@
+﻿# Lambda improvements
+
+## Summary
+Proposed changes:
+1. Allow lambdas with attributes
+2. Allow lambdas with explicit return type
+3. Infer a natural delegate type for lambdas and method groups
+
+## Motivation
+Support for attributes on lambdas would provide parity with methods and local functions.
+
+Support for explicit return types would provide symmetry with lambda parameters where explicit types can be specified.
+Allowing explicit return types would also provide control over compiler performance in nested lambdas where overload resolution must bind the lambda body currently to determine the signature.
+
+A natural type for lambda expressions and method groups will allow more scenarios where lambdas and method groups may be used without an explicit delegate type, including as initializers in `var` declarations.
+
+Requiring explicit delegate types for lambdas and method groups has been a friction point for customers, and has become an impediment to progress in ASP.NET with recent work on [MapAction](https://github.com/dotnet/aspnetcore/pull/29878).
+
+[ASP.NET MapAction](https://github.com/dotnet/aspnetcore/pull/29878) without proposed changes (`MapAction()` takes a `System.Delegate` argument):
+```csharp
+[HttpGet("/")] Todo GetTodo() => new(Id: 0, Name: "Name");
```

Comment:
@halter73 we should get these samples updated.

Reply:
Yeah. One issue with that is the current updated samples rely heavily on conventions for parameter binding and no longer use an attribute for specifying the route pattern or HTTP method. That said, all that parameter attributes can still be optionally applied.

```C#
app.MapPut("/todos/{id}", async ([FromRoute(Name = "id")] int identifier, [FromBody] Todo inputTodo, [FromServices] TodoDb db) =>
{
    var todo = await db.Todos.FindAsync(identifier);

    if (todo is null) return NotFound();

    todo.Title = inputTodo.Title;
    todo.IsComplete = inputTodo.IsComplete;

    await db.SaveChanges();

    return NoContent();
});
```